### PR TITLE
Best practices for installing software

### PR DESCRIPTION
**Pull request by seisman**
Perhaps we can also incorporate the content of https://blog.seisman.info/how-to-install-softwares-under-centos-7/ into this section, as how to and where to install packages are related questions.

BTW, this section will be much shorter than the original post, as Fedora doesn't have the many problems as CentOS.

TODO:

- [ ] Maybe also mention vcpkg

### CODE REVIEW

File: source/best-practices/installation-path.rst [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+到 ``~/opt/`` 的子目录下，比如 ``~/opt/TauP`` 和 ``~/opt/GMT``\ 。
+
+~/bin
+-----
+
+对于简单的小工具，比如单个源码文件或单个可执行文件，可以直接复制到 ``~/bin`` 目录，
+如 ``distaz``
+
+/opt
+----
```
Comment (core-man) [inline source/best-practices/installation-path.rst:35]: Maybe it's better to classify each subsection by the type of software. E.g.,

- 软件包
- 无需编译软件
- 编译软件
- 闭源软件

Comment (seisman) [reply source/best-practices/installation-path.rst:35]: Yes, will do it tomorrow, or you can work in it.

Comment (core-man) [reply source/best-practices/installation-path.rst]: Revised~

**Comment by core-man** [comment]
1. if source code needs to be compiled but the bin files are not in the bin folder, e.g. fk. do we want to suggest to make one.
2. do we want to add install path by some package managers, e.g., apt-get，dng, conda, homebrew, macports. actually, we don't need to care about it.

**Comment by seisman** [comment]
> 2\. do we want to add install path by some package managers, e.g., apt-get，dng, conda, homebrew, macports. actually, we don't need to care about it.

No.

File: source/best-practices/software-installation.rst [hunk 2]
```diff
@@ -1,6 +1,32 @@
 软件安装最佳实践
 ================
 
+软件安装方式
+------------
+
+Linux 或 macOS 系统下安装软件通常有如下几种方式：
+
+1. 使用系统自带的应用商店（App Store）在图形界面下安装软件
```
Comment (core-man) [inline source/best-practices/software-installation.rst:9]: This only work for macOS?

Comment (seisman) [reply source/best-practices/software-installation.rst:9]: > This only work for macOS?

Linux also has App Stores, but unlike macOS, installing a software from the Linux App Store is the same as running `apt` or `yum`. Linux App Store is just a GUI of apt/yum.

File: source/best-practices/software-installation.rst [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+:预计阅读时间: X 分钟
+
+软件安装方式
+------------
+
+Linux 或 macOS 系统下安装软件通常有如下几种方式：
+
+1. 使用系统自带的应用商店（App Store）在图形界面下安装软件
+2. 使用软件包管理器（如 Ubuntu/Debian 的 ``apt``、Fedora 的 ``dnf``，
+   或 macOS 的 ``brew``）通过命令行安装官方源提供的软件
```
Comment (core-man) [inline source/best-practices/software-installation.rst:15]:
```suggestion
2. 使用软件包管理器通过命令行安装软件。例如，Ubuntu/Debian 的 ``apt``\ 、
    Fedora 的 ``dnf``\ 、macOS 的 ``brew``\ 。这些软件包管理器既可以安装
    官方源提供的软件，也可以安装第三方源提供的软件。最好选择可信赖
    的第三方源。
```

File: source/best-practices/software-installation.rst [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+3. 使用软件包管理器通过命令行安装第三方源提供的软件
```
Comment (core-man) [inline source/best-practices/software-installation.rst:16]:
```suggestion
```

File: source/best-practices/software-installation.rst [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+4. 直接下载闭源软件或商业软件提供的二进制包（如 Linux 下的 deb 和 rpm，macOS 下的 dmg）
+   进行安装
+5. 某个编程语言的包，通常可以使用该编程语言自带的包管理工具进行安装，比如
+   Python 编程语言的包管理器 ``conda`` 和 ``pip``
+6. 某些软件提供了二进制的压缩包，直接解压放在某个目录，然后修改环境变量
+   **PATH** 即可使用
+7. 下载软件源码、编译并安装
```
Comment (core-man) [inline source/best-practices/software-installation.rst:23]:
```suggestion
3. 直接下载闭源软件或商业软件提供的二进制包进行安装，如 Linux 下的 ``deb`` 和 ``rpm``\ 、
   macOS 下的 ``dmg``\ 。
4. 某个编程语言的包，通常可以使用该编程语言自带的包管理工具进行安装，比如
   Python 语言的包管理器 ``conda`` 和 ``pip``
5. 某些软件提供了二进制的压缩包，则解压后直接放在某个目录，然后修改环境变量
   **PATH** 即可使用
6. 下载软件源码、编译并安装
```

File: source/best-practices/software-installation.rst [hunk 6]
```diff
@@ -0,0 +1,93 @@
+软件安装最佳实践
+================
+
+:本节贡献者: `姚家园 <https://github.com/core-man>`__ （作者）、`田冬冬 <https://me.seisman.info>`__ （作者）
+:最近更新日期: 2021-01-12
+:预计阅读时间: X 分钟
+
+软件安装方式
+------------
```
Comment (core-man) [inline source/best-practices/software-installation.rst:9]:
```suggestion
安装方式
--------
```

File: source/best-practices/software-installation.rst [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+都可以通过方式 1—4 和方式 6 安装；而对于地震学科研软件而言，
+则可以通过方式 5（很多地震学软件是用 Python 语言写的，因而可以通过 ``conda``
+或 ``pip`` 安装）、方式 6 或方式 7 进行安装。
+
+方式 1—5 会将软件安装到默认的位置，通常用户无法控制也无需关心其具体安装位置。
+而对于方式 6 和 7，用户可选择将软件安装到任意位置。为了便于安装与管理科研
+相关软件，我们总结了如下最佳实践。
+
+软件安装路径
+------------
```
Comment (core-man) [inline source/best-practices/software-installation.rst:36]:
```suggestion
安装路径
--------
```

File: source/best-practices/software-installation.rst [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+6. 下载软件源码、编译并安装
+
+绝大多数非科研类开源软件都可以通过方式 1—3 安装；多数商业软件
+都可以通过方式 1—4 和方式 6 安装；而对于地震学科研软件而言，
+则可以通过方式 5（很多地震学软件是用 Python 语言写的，因而可以通过 ``conda``
+或 ``pip`` 安装）、方式 6 或方式 7 进行安装。
+
+方式 1—5 会将软件安装到默认的位置，通常用户无法控制也无需关心其具体安装位置。
+而对于方式 6 和 7，用户可选择将软件安装到任意位置。为了便于安装与管理科研
+相关软件，我们总结了如下最佳实践。
```
Comment (core-man) [inline source/best-practices/software-installation.rst]:
```suggestion
方式 1—4 会将软件安装到默认的位置，通常用户无法控制，也无需关心其具体安装位置。
而对于方式 5 和 6，用户可选择将软件安装到任意位置。为了便于安装与管理科研
相关软件，我们总结了如下最佳实践。
```

File: source/best-practices/software-installation.rst [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+绝大多数非科研类开源软件都可以通过方式 1—3 安装；多数商业软件
+都可以通过方式 1—3 和方式 5 安装；而对于地震学科研软件而言，
+则可以通过方式 4（很多地震学软件是用 Python 语言写的，因而可以通过 ``conda``
+或 ``pip`` 安装）、方式 5 或方式 6 进行安装。
+
+方式 1—4 会将软件安装到默认的位置，通常用户无法控制也无需关心其具体安装位置。
+而对于方式 5 和 6，用户可选择将软件安装到任意位置。为了便于安装与管理科研
+相关软件，我们总结了如下最佳实践。
+
```
Comment (core-man) [inline source/best-practices/software-installation.rst:34]:
```suggestion
方式 1—4 会将软件安装到默认的位置，通常用户无法控制，也无需关心其具体安装位置。
而对于方式 5 和 6，用户可选择将软件安装到任意位置。为了便于安装与管理科研
相关软件，我们总结了如下最佳实践。

```

File: source/best-practices/software-installation.rst [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+4.  某个编程语言的包，通常可以使用该编程语言自带的包管理工具进行安装，比如
+    Python 语言的包管理器 ``conda`` 和 ``pip``
+5.  某些软件提供了二进制的压缩包，则解压后直接放在某个目录，然后修改环境变量
+    **PATH** 即可使用
+6.  下载软件源码、编译并安装
+
+绝大多数非科研类开源软件都可以通过方式 1—3 安装；多数商业软件
+都可以通过方式 1—3 和方式 5 安装；而对于地震学科研软件而言，
+则可以通过方式 4（很多地震学软件是用 Python 语言写的，因而可以通过 ``conda``
+或 ``pip`` 安装）、方式 5 或方式 6 进行安装。
```
Comment (core-man) [inline source/best-practices/software-installation.rst:29]:
```suggestion
绝大多数非科研类开源软件都可以通过方式 1—3 安装；多数闭源软件或商业软件
都可以通过方式 1—3 和方式 5 安装；对于地震学科研软件而言，
通常可以通过方式 4（很多地震学软件是用 Python 语言写的，因而可以通过 ``conda``
或 ``pip`` 安装）、方式 5 或方式 6 进行安装。
```

File: source/best-practices/software-installation.rst [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+从源码编译的软件
+^^^^^^^^^^^^^^^^
+
+需要从源码编译的软件，可将源码目录放在 :file:`~/src/` 的子目录下，如 :file:`~/src/fk`\ 、
+:file:`~/src/GMT`\ 、:file:`~/src/distaz`\ 。
+
+若该软件不需安装，则编译链接后生成的可执行文件在该软件目录下或其子目录 :file:`bin` 下，
+如 :file:`~/src/fk`\ 、:file:`~/src/CPS/bin`\ 。如果该软件只是一个简单的小工具，
+比如单个源码文件或单个可执行文件，则可以将生成的可执行文件复制至 :file:`~/bin/` 目录下，
+如 :file:`~/bin/distaz`\ 。
```
Comment (core-man) [inline source/best-practices/software-installation.rst:67]: How do we deal with fk? its binary files are in the source dir.

File: source/best-practices/software-installation.rst [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+方式 1—4 会将软件安装到默认的位置，通常用户无法控制也无需关心其具体安装位置。
+而对于方式 5 和 6，用户可选择将软件安装到任意位置。为了便于安装与管理科研
+相关软件，我们总结了如下最佳实践。
+
+软件安装路径
+------------
+
+软件安装相关的路径如下：
+
+- :file:`~/software/`
```
Comment (core-man) [inline source/best-practices/software-installation.rst:40]:
```suggestion
- :file:`~/software/`\ ：存放软件的原始压缩包
```

File: source/best-practices/software-installation.rst [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+- :file:`~/opt/`
+- :file:`~/src/`
+- :file:`~/bin/`
```
Comment (core-man) [inline source/best-practices/software-installation.rst:43]:
```suggestion
- :file:`~/bin/`\ ：简单的小工具软件，如单个源码文件生成的可执行文件或单个可执行文件
```

File: source/best-practices/software-installation.rst [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+相关软件，我们总结了如下最佳实践。
+
+软件安装路径
+------------
+
+软件安装相关的路径如下：
+
+- :file:`~/software/`
+- :file:`~/opt/`
+- :file:`~/src/`
```
Comment (core-man) [inline source/best-practices/software-installation.rst:42]:
```suggestion
- :file:`~/src/`\ ：需源码编译的软件（含需安装软件的源码目录）、
```

File: source/best-practices/software-installation.rst [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+而对于方式 5 和 6，用户可选择将软件安装到任意位置。为了便于安装与管理科研
+相关软件，我们总结了如下最佳实践。
+
+软件安装路径
+------------
+
+软件安装相关的路径如下：
+
+- :file:`~/software/`
+- :file:`~/opt/`
```
Comment (core-man) [inline source/best-practices/software-installation.rst:41]:
```suggestion
- :file:`~/opt/`\ ：解压即用的软件、需源码编译并安装的软件
```

File: source/best-practices/software-installation.rst [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+- :file:`~/src/`
+- :file:`~/bin/`
+- :file:`/opt/`
```
Comment (core-man) [inline source/best-practices/software-installation.rst:44]:
```suggestion
- :file:`/opt/`\ ：大多数闭源软件或商业软件、某些与地震学科研无关的软件
```

File: source/best-practices/software-installation.rst (hunk 11 above)
Comment (seisman) [reply source/best-practices/software-installation.rst:67]: Considering that fk provides several commands and scripts, perhaps the easiest way is still to add the fk directory to PATH

File: source/best-practices/software-installation.rst [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+
+安装路径
+--------
+
+软件安装相关的路径如下：
+
+- :file:`~/software/`\ ：存放软件的原始压缩包
+- :file:`~/opt/`\ ：解压即用的软件、需源码编译并安装的软件
+- :file:`~/src/`\ ：需源码编译的软件（含需安装软件的源码目录）
+- :file:`~/bin/`\ ：简单的小工具软件，如单个源码文件生成的可执行文件或单个可执行文件
```
Comment (core-man) [inline source/best-practices/software-installation.rst:43]:
```suggestion
- :file:`~/bin/`\ ：简单的小工具软件，如单个源码文件编译链接后生成的可执行文件、单个可执行文件
```

File: source/best-practices/software-installation.rst [hunk 18]
```diff
@@ -0,0 +1,94 @@
+软件安装最佳实践
+================
+
+:本节贡献者: `姚家园 <https://github.com/core-man>`__ （作者）、`田冬冬 <https://me.seisman.info>`__ （作者）
+:最近更新日期: 2021-01-12
+:预计阅读时间: X 分钟
```
Comment (seisman) [inline source/best-practices/software-installation.rst:6]:
```suggestion
:预计阅读时间: 10 分钟
```

File: source/best-practices/software-installation.rst (hunk 17 above)
Comment (seisman) [reply source/best-practices/software-installation.rst:43]: "以及工具型脚本"？

File: source/best-practices/software-installation.rst (hunk 11 above)
Comment (core-man) [reply source/best-practices/software-installation.rst:67]: Shall we make a bin folder under ``fk` instead? although this is not related to the topic here.

Comment (seisman) [reply source/best-practices/software-installation.rst:67]: Maybe no.

File: source/best-practices/software-installation.rst [hunk 19]
```diff
@@ -0,0 +1,94 @@
+软件安装最佳实践
+================
+
+:本节贡献者: `姚家园 <https://github.com/core-man>`__ （作者）、`田冬冬 <https://me.seisman.info>`__ （作者）
```
Comment (core-man) [inline source/best-practices/software-installation.rst:4]:
```suggestion
:本节贡献者: `田冬冬 <https://me.seisman.info>`__ （作者）、`姚家园 <https://github.com/core-man>`__ （作者）
```

File: source/best-practices/software-installation.rst (hunk 17 above)
Comment (core-man) [reply source/best-practices/software-installation.rst:43]: Yes

File: source/best-practices/software-installation.rst [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+相关软件，我们总结了如下最佳实践。
+
+安装路径
+--------
+
+软件安装相关的路径如下：
+
+- :file:`~/software/`\ ：存放软件的原始压缩包
+- :file:`~/opt/`\ ：解压即用的软件、需源码编译并安装的软件
+- :file:`~/src/`\ ：需源码编译的软件（含需安装软件的源码目录）
```
Comment (core-man) [inline source/best-practices/software-installation.rst:42]:
```suggestion
- :file:`~/src/`\ ：需源码编译的软件（含需源码并安装软件的源码目录）
```

File: source/best-practices/software-installation.rst [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+- :file:`~/bin/`\ ：简单的小工具软件，如单个源码文件编译链接后生成的可执行文件、
+  单个可执行文件以及工具型脚本
+- :file:`/opt/`\ ：大多数闭源软件或商业软件、某些与地震学科研无关的软件
+
+软件包
+^^^^^^^
+
+软件的原始压缩包可以保存在 :file:`~/software/` 目录下, 相当于做一个备份，
```
Comment (core-man) [inline source/best-practices/software-installation.rst:50]:
```suggestion
软件的原始压缩包可以保存在 :file:`~/software/` 目录下，相当于做一个备份，
```

File: source/best-practices/software-installation.rst (hunk 18 above)
Comment (core-man) [reply source/best-practices/software-installation.rst:6]: Almost 10 mins.

**Review by core-man — APPROVED** [review]

File: source/best-practices/software-installation.rst [hunk 22]
```diff
@@ -0,0 +1,95 @@
+软件安装最佳实践
+================
```
Comment (core-man) [inline source/best-practices/software-installation.rst:2]: Remove 最佳实践?

Comment (seisman) [reply source/best-practices/software-installation.rst:2]: OK to me.

Comment (seisman) [inline source/best-practices/software-installation.rst:2]:
```suggestion
软件安装
========
```

Comment (seisman) [reply source/best-practices/software-installation.rst]: Done.

**Review by core-man — APPROVED** [review]

**Pull request merged**